client/network/src/request_responses.rs

```diff
@@ -121,23 +121,34 @@ pub enum Event {
 	/// A remote sent a request and either we have successfully answered it or an error happened.
 	///
 	/// This event is generated for statistics purposes.
+	///
+	/// Note: Multiple [`InboundRequest`] events can be emitted for the same inbound request. E.g.
+	/// when response building succeeds, but sending the response later on fails. In that case there
+	/// will be one [`InboundRequest`] with the time it took to build the response and one
+	/// [`InboundRequest`] with the reason why later on sending the response failed.
 	InboundRequest {
 		/// Peer which has emitted the request.
 		peer: PeerId,
 		/// Name of the protocol in question.
 		protocol: Cow<'static, str>,
 		/// If `Ok`, contains the time elapsed between when we received the request and when we
-		/// sent back the response. If `Err`, the error that happened.
+		/// started sending the response. If `Err`, the error that happened.
 		result: Result<Duration, ResponseFailure>,
 	},
```

Comment:
> Multiple [`InboundRequest`] events can be emitted for the same inbound request.

This really looks like a hack to me.

Is there no way to report only one event per request, and indicate both the duration and whether the sending has succeeded in the same event?

Reply:
Yes, this hack would be removed with https://github.com/libp2p/rust-libp2p/pull/1867.